### PR TITLE
toloka/rutracker: improve parsing and parsing titles

### PR DESCRIPTION
**Pull request by mynameisbogdan**
List of indexers with `Split('=')` usage:

- [x] toloka
- [x] norbits: moved to #13955
- [ ] bjshare
- [x] funfile: moved to  #13954
- [x] pornolab: moved to #13956
- [x] pretome: moved to #13957
- [x] rutracker
- [ ] torrentheaven
- [ ] zonaq

### CODE REVIEW

**Comment by garfield69** [comment]
>norbits: I wonder why the link and details was build instead of being taken from the page?

norbits is one of the oldest indexers first introduced 6 years ago by the author when he birthed Jackett.
going back in history tracks I see that details and download links were being built right back then at the start, so presumably the web site did not provide those html selectors as hardcoded tags, and I suppose it most likely that they were dynamically build using JS at browser run time, and so  Jackett did not have access to them. maybe, possibly, wtfk ;-D
what is true is that no one has ever reviewed the indexer since and brought it up to date with the  current web  site support.
I myself, am not a C# dev, specialising on the yaml side of the fence, so I leave the C# indexers mostly alone, only hacking simple fixes and updates when forced to do so.

File: src/Jackett.Common/Indexers/RuTracker.cs [hunk 1]
```diff
@@ -1606,23 +1609,18 @@ private ReleaseInfo ParseReleaseRow(IElement row)
...
-                {
                     release.Title = MoveFirstTagsToEndOfReleaseTitle(release.Title);
-                }
-
                 if (release.Category.Contains(TorznabCatType.Audio.ID))
-                {
                     release.Title = DetectRereleaseInReleaseTitle(release.Title);
-                }
+
+                release.Title = release.Title.Trim(' ', '.', '-', '_', '|', '/', '\'');
```
Comment (ilike2burnthing) [inline src/Jackett.Common/Indexers/RuTracker.cs:1623]:
```suggestion
                release.Title = release.Title.Trim(' ', '.', '-', '_', '|', '/', '\');
```

**Review by ilike2burnthing — CHANGES_REQUESTED** [review]
Looks good, I mostly just found hang overs from the old indexers.

Speaking of, rutracker's `captchaimg` selector URL (L1392) needs changed to `https://static.rutracker.cc/captcha/`.

File: src/Jackett.Common/Indexers/Toloka.cs [hunk 2]
```diff
@@ -344,5 +317,31 @@ protected override async Task<IEnumerable<ReleaseInfo>> PerformQuery(TorznabQuer
 
             return releases;
         }
+
+        private static string CleanTitle(string title, ICollection<int> category, bool stripCyrillicLetters = true)
+        {
+            var tvShowTitleRegex = new Regex(".+\\/\\s([^а-яА-я\\/]+)\\s\\/.+Сезон\\s*[:]*\\s+(\\d+).+(?:Серії|Епізод)+\\s*[:]*\\s+(\\d+-*\\d*).+,\\s+(.+)\\]\\s(.+)", RegexOptions.Compiled | RegexOptions.IgnoreCase);
```
[the remark this answers is not in the document]
Comment (ilike2burnthing) [reply src/Jackett.Common/Indexers/Toloka.cs:323]: Oh, my bad, I didn't scroll far enough to the right and was instead just looking for `(?i)`. I'll have a closer look at the issues I was finding, as I thought I'd pinned it down to the case, but obviously not.

[the remark this answers is not in the document]
Comment (ilike2burnthing) [reply src/Jackett.Common/Indexers/Toloka.cs:323]: You read my mind, I have https://regexr.com open at the moment. Will report back shortly.

**Review by ilike2burnthing — CHANGES_REQUESTED** [review]
So the problem was that it was overcomplicated, excluding a lot of titles. So my proposed solution is to overcomplicate it in a different way ;) Not tested in Jackett.

Equivalent for rutracker (replaces L1579-L1584):

```C#
                    var regex_SEof = new Regex("Сезон\\s*[:]*\\s+(\\d+).+(?:Серии|Эпизод)+\\s*[:]*\\s+(\\d+-*\\d*)\\s*из\\s*(\\w+)");
                    var regex_SE = new Regex("Сезон\\s*[:]*\\s+(\\d+).+(?:Серии|Эпизод)+\\s*[:]*\\s+(\\d+-*\\d*)");
                    var regex_S = new Regex("Сезон\\s*[:]*\\s+(\\d+)");
                    var regex_Eof = new Regex("(?:Серии|Эпизод)+\\s*[:]*\\s+(\\d+-*\\d*)\\s*из\\s*(\\w+)");
                    var regex_E = new Regex("(?:Серии|Эпизод)+\\s*[:]*\\s+(\\d+-*\\d*)*(\\w+)");

                    //replace double 4K quality in title
                    release.Title = release.Title.Replace(", 4K]", "]");

                    var title = regex_SEof.Replace(release.Title, "S$1E$2 of $3");
                    var title = regex_SE.Replace(release.Title, "S$1E$2");
                    var title = regex_S.Replace(release.Title, "S$1");
                    var title = regex_Eof.Replace(release.Title, "E$1 of $2");
                    var title = regex_E.Replace(release.Title, "E$1");
```

Make changes as needed, I'm not precious about 'my' copied and pasted code.

**Review by ilike2burnthing — COMMENTED** [review]
There's been a minor regression with Toloka, as I'm seeing the inconsistent results with `shrinking webrip` (and others) again.

Everything else looks great!

**Review by ilike2burnthing — CHANGES_REQUESTED** [review]
Right! Hopefully the last review from me, as I think I've found the problem:
use `IsAnyTvCategory` from RuTracker, rather than
`if (category.Contains(TorznabCatType.TV.ID))`, as it was only applying to `TV` results and not subcategories.

You can revert the last commit, as this was then removing any season and episode numbers from results if `stripCyrillic` was enabled.

**Comment by garfield69** [comment]
in general, the series name should go first. some *arr apps will not match the title.

**Review by ilike2burnthing — CHANGES_REQUESTED** [review]
Suggested changes to prevent results like `(S1E4-​)` and `(E01-04 of ​)` due to the use of the Cyrillic `Х`, e.g. `серії 1-4 з ХХ`. Also a couple of extra characters to trim.

As for the issue of the season and episodes being left at the start of the title when Cyrillic characters are stripped, can we use something like `MoveFirstTagsToEndOfReleaseTitle` in RuTracker to resolve this (without the need to make it a user setting)?

**Comment by garfield69** [comment]
So, I've lost track of what's what atm. are the new pulls that this pull spawned, indexers that are final are ready for one last review by yours truly?

**Comment by ilike2burnthing** [comment]
Today is a bit of a mess, so will have to test these later (ah balls, I've somehow made it worse). Thanks for sticking with the million revisions :P 

For `_tvTitleEngSeasonEpisodeOfRegex` and `_tvTitleEngSeasonEpisodeRegex` in Toloka, why was `(\d+(?:-\d+)?)` used for the episodes instead of `(\d+-*\d*)`?

**Comment by ilike2burnthing** [comment]
Remaining issues according to the test failures:
- the rutracker test is failing with `System.Reflection.TargetParameterCountException : Parameter count mismatch`
- when `stripCyrillicLetters` is true, the (previously) Cyrillic season and episode numbers are removed completely, which then means `moveFirstTagsToEndOfReleaseTitle` is instead moving the English season and episode numbers
- The start of `(S1-2, E14 of 20) / She-Ra and the Princesses of Power (S1-2, E14-20 of 20) (2018) WEBRip 1080p` is not recognised as a tag

These are all beyond me, so I pass the reins back to you.

Also, both regex101.com and regexr.com don't like `(?<-c>)` from Toloka's `findTagsInTitlesRegexList`. Is this just an issue with those sites, or the code?

**Comment by ilike2burnthing** [comment]
> Did you selected the .NET (C#) flavor?

Well that would just be smart :P 
Yea, now it's happy.

**Comment by ilike2burnthing** [comment]
No worries lol

Also, complete aside, I just realised that currently Jackett is being maintained by 3 cartoon characters. Seems normal 😉

**Comment by ilike2burnthing** [comment]
You didn't miss much, just `_tvTitleCommaRegex` and `_tvTitleCyrillicXRegex`, added your `(\d+(?:-\d+)?)` where needed (and fixed the replacements for them), and then a few tweaks to the tests.

The tests look good, bar that one oddity. I'll expand them later.

**Review by ilike2burnthing — APPROVED** [review]
Looks good!

There are some other minor variations that aren't covered by all of this, but I stopped caring a while ago :P

**Comment by ilike2burnthing** [comment]
I'm happy as is, so @garfield69 can take a look and merge when they're free.

**Review by garfield69 — APPROVED** [review]

**Comment by ilike2burnthing** [comment]
@mynameisbogdan & @garfield69 would it be worth adding a `stripcyrillic` or `striprussian` filter to the Cardigann indexer using `\p{IsCyrillic}`? It would need to be boolean so we could still use something like:

```yml
title:
  selector: blah
  filters:
    - name: stripcyrillic
      args: "{{ if .Config.striprussian }}true{{ else }}false{{ end }}"
```

Or is there a better way to implement that?

**Comment by ilike2burnthing** [comment]
Sorry, I was meaning here:

https://github.com/Jackett/Jackett/blob/3adf75097315b23bef447a534c7cb6278223a8e5/src/Jackett.Common/Indexers/CardigannIndexer.cs#L996

**Comment by ilike2burnthing** [comment]
Ohhhhh, gotcha! I assumed it didn't, but let me check.

**Comment by ilike2burnthing** [comment]
Trying `'["[\p{IsCyrillic}\\/]+", ""]'` in rutor.yml results in the error:

```text
System.Exception: Error while parsing field=title, selector=td:nth-of-type(2) a[href^="/torrent/"], value=<null>: Cannot convert type 'char' to 'string'
   at Jackett.Common.Indexers.CardigannIndexer.PerformQuery(TorznabQuery query) in D:\a\1\s\src\Jackett.Common\Indexers\CardigannIndexer.cs:line 1638
```

Double external quotes
`While parsing a block mapping, did not find expected key.`
No external quotes
`While scanning a quoted scalar, found unknown escape character.`
Double external quotes and single internal quotes
`While scanning a quoted scalar, found unknown escape character.`

**Comment by ilike2burnthing** [comment]
Oh ffs, that's what I get for trying to do 5 things at once. Ta!

**Pull request merged**